lib/Sema/SemaBounds.cpp

```diff
@@ -896,10 +1129,15 @@ namespace {
       // Bounds of the right-hand side of the assignment
       BoundsExpr *RHSBounds = nullptr;
 
-      // Check that the value being assigned has bounds if the
-      // target of the LHS lvalue has bounds.
-      if (LHSType->isCheckedPointerType() ||
+      if (!E->isCompoundAssignmentOp() &&
+          LHSType->isCheckedPointerPtrType() &&
+          RHS->getType()->isCheckedPointerPtrType()) {
+        // ptr<T> to ptr<T> assignment, no obligation to infer any bounds for either side
```

Comment:
This code is new - it stops requiring bounds for both sides if it is a `ptr<T>` to `ptr<T>` assignment.